### PR TITLE
Upgrade command

### LINKED ISSUE #61: Add `exercism upgrade` or `exercism update`
...

### PR DESCRIPTION
**Pull request by Tonkpils**
This is a working prototype of the `upgrade` command for the CLI.

It will install the latest version of the CLI under the same path as the executable that called the command. 

One thing that it does not support yet that I'd like to fix before this gets merged:

1. **ARM version upgrades**
  Golang has no way of checking the ARM version during runtime
    - A possible solution would be to set the ARM version during the build process with `-ldflags` which would allow us to set the binary's OS/ARCH/ARMV that was used during build. 

Also, I wrote the windows portion a bit blindly since I don't have a windows machine to test it on. I'll make sure it works properly when I get a chance. 

I'd also like to include a generic message to file an issue here if the install fails with the `debug` info but not too sure how the message should read. Thoughts?

Once this is merged it fixes #61

### CODE REVIEW

**Comment by kytrinyx** [comment]
This is so cool! I'm at work, so can't read through it now. I think setting the arm version with a build flag is perfect.

**Comment by Tonkpils** [comment]
Ok, I re-wrote the `bin/build-all` script to make it set the variables during build. I've tested the upgrade command in linux, windows and mac. Also, if you see any improvements that can be done to the build script let me know.

File: api/api.go [hunk 1]
```diff
@@ -266,6 +266,7 @@ func (c *Client) Tracks() ([]*Track, error) {
 	return payload.Tracks, nil
 }
 
+// Skip sends a request to exercism to skip the exercise given language and slug
```
Comment (kytrinyx) [inline api/api.go]: Yay, doc comments :)

File: bin/build-all [hunk 2]
```diff
@@ -5,51 +5,75 @@ set -e -x
...
+createRelease linux 386
+createRelease linux amd64
+
+# ARM Releases
+createRelease linux arm 5
+createRelease linux arm 6
+
+# Windows Releases
+createRelease windows 386
+createRelease windows amd64
```
Comment (kytrinyx) [inline bin/build-all]: This whole thing is so much nicer!

File: cmd/upgrade.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+// Upgrade command allows the user to upgrade to the latest CLI version
+func Upgrade(ctx *cli.Context) {
+	client := http.Client{Timeout: 5 * time.Second}
+	rel, err := checkLatestRelease(client)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// TODO: This checks strings against string
+	// Version 2.2.0 against release 2.1.0
```
Comment (kytrinyx) [inline cmd/upgrade.go]: The TODO comment is a bit unclear. Do you mean that we should check semver and only auto upgrade minor or patch versions automatically?

File: cmd/upgrade.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+	// will trigger an upgrade...
+	// should probably parse semver and compare
+	if rel.Version() == ctx.App.Version {
+		fmt.Println("Your CLI is up to date!")
+		return
+	}
+
+	// current executable path
+	dest, err := osext.Executable()
```
Comment (kytrinyx) [inline cmd/upgrade.go]: I was wondering how we would do that. Very nice!

File: cmd/upgrade.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+type asset struct {
+	ID          int    `json:"id"`
+	Name        string `json:"name"`
+	ContentType string `json:"content_type"`
+}
+
+func (a *asset) Download() (*bytes.Reader, error) {
+	downloadURL := fmt.Sprintf("https://api.github.com/repos/exercism/cli/releases/assets/%d", a.ID)
```
Comment (kytrinyx) [inline cmd/upgrade.go]: Oh, wow. This is way better than what I'm doing in the bash install script (which mucks around with location headers).

**Comment by kytrinyx** [comment]
This looks great. I'm going to merge it in (the only thing that I was wondering about is a comment, and that can be tweaked later, I think).

**Pull request merged**